Adjustments/Additions for Better Compatibility & Performance

ICY105 opened the pull request:
I plan on integrating CS into Datapack Utils moving into 1.15. With that in mind, I made 3 primary changes to CS. Feel free to review them (maybe we can get DU/GM4 compatability):

1.  Removed the load function. There are cases where CS should not run when it is installed (for example, Datapack Utils before another datapack hooks into its API). CS can be enabled by scheduling its main function as needed.

2. Added chunk spreading to the End using moving piston blocks at y=255.

3. Altered the spreading to fill barriers/moving pistons as an entity is created instead of when generated, circumventing the need to entity checks as it spreads. The downside is if the AEC gets killed somehow, the chunk(s) will not receive any generation.

ICY105 committed:
Removed auto load. Allows CS to remain disabled if needed.

Some packs (like Datapack Utils) shouldn't run CS on it's own, but only when another pack is added. This change allows datapacks to have CS but only run it when need (manually activated by scheduling main function).

ICY105 committed:
Barriers are filled as the AEC is summoned.

Completely prevents overlapinig AEC generation. Avoid needless @e checks.

ICY105 committed:
Adds chunk spreading to The End using moving piston blocks at y=255

ICY105 commented:
After some testing I've realized that removing the load function still makes main run after scheduling it for the first time, so a better solution is needed- potentially use Lantern for guaranteed load order?

Also, AEC are not spread into unloaded chunks, making the cut off in generation very abrupt. A new change elevates that somewhat by re-starting generation 6 chunks away. Most noticeable on higher render distances.

ICY105 committed:
Added furthure spread chunks.

AEC are not spread into unloaded chunks, meaning a player has to move all the way onto an ungenerated chunk for generation to resume. This counters that somewhat by starting the spread 6 chunks away, when applicable. Most noticable on high render distances.

ICY105 committed:
Added LanternMC to control load order and versioning

ICY105 commented:
Alright, should be the last change. Added LanternMC so other datapacks can enable CS in post load without having to worry about load order. They can also check the version to test for conflicts with outdated datapacks.

All the the major changes are separated into different commits, so you should be able to sort through the ones you want to keep/discard if you don't want all the changes.

misode commented:
> Added LanternMC Load

Thanks, good idea

> There are cases where CS should not run when it is installed

The point of CS is to be a universal chunk scan system. You can't rely on it not running or only running when you want it. The only thing data packs should do is respond to the function tag call.

> Altered the spreading to fill barriers/moving pistons as an entity is created instead of when generated, circumventing the need to entity checks as it spreads. The downside is if the AEC gets killed somehow, the chunk(s) will not receive any generation.

I would consider this to be a pretty major downside. 

> Added further spread chunks

Why is this necessary? The spreading will happen automatically so this should not make any difference. Except for adding additional commands.

> Added chunk spreading to the End
 
Looks good

ICY105 commented:
`There are cases where CS should not run when it is installed`

My reasoning for this is that Datapack Utils should not enable chunk spreader until another datapack adds world gen (for example Mech adds custom ores). Otherwise the spreader is running but generates nothing. However, this is an edge case that would rarely be used, and the side effects when it does are minor. It also increases the burden on the average user since they have to enable generation, so I can see this being reverted.

`Altered the spreading to fill barriers/moving pistons as an entity is created instead of when generated, circumventing the need to entity checks as it spreads. The downside is if the AEC gets killed somehow, the chunk(s) will not receive any generation.`

This is a performance/side effect trade off. I'll leave it up to you what to do with it. _Although as I'm writing this I think I may have a hybrid solution that has slightly better performance without the side effect..._ 

`Added further spread chunks.`

Strictly speaking, it's not necessary.  It just makes generation appear smoother with less stuff popping up in your face. AECs don't spread into unloaded chunks, so it's possible to have a sharp cutoff at your render distance. Generation will resume, but not until you've moved all the way into the ungenerated area (it's sort of a hybrid between DU's super chunk system and entity spreading). This change makes it so generation will resume 6 chunks away so it's less abrupt.

misode commented:
Not automatically running can cause problems, for example when existing chunks that players have built in, get overwritten. If you install a (new) terrain modification data pack, it should only alter newly generated chunks.

If you don't want the Datapack Utils core to start the chunk scanning, just don't bundle this data pack with it.

---

> Generation will resume, but not until you've moved all the way into the ungenerated area

I don't think that's true... If you move just one chunk in a new direction, all the chunks at the render distance will be generated, since the area effect cloud is there.

ICY105 commented:
`Generation will resume, but not until you've moved all the way into the ungenerated area`

The problem is this:

`execute positioned ~16 ~ ~ <if block ~ ~ ~ minecraft:bedrock> run summon minecraft:area_effect_cloud ~ ~ ~ {Tags:[cs.chunk],Duration:2147483647}`

What happens at the render distance is an AEC will run generation and kill itself, but will not summon a new AEC outside the render distance because `if block` returns `position is not loaded`. Or at least, that's what I've been noticing in my test world.

Here's a screenshot without the change (at render distance 12). After standing still for about a minute, this occurred when moving into a new area:

![Imgur](https://i.imgur.com/eKCLeCK.png)

ICY105 committed:
Reverted change to loading. Main runs immediatly as before.

misode commented:
Okay I see the issue. In GM4 this was solved by adding a `distance=..100` check. Would you consider this a valid solution?

ICY105 commented:
Still has a chance to occur for smaller render distances (< 8), but I don't think it would have a noticeable impact on their gameplay since stuff would be popping up in their face all the time.

ICY105 committed:
Removed "further spreading", added distance cap.

ICY105 committed:
Reverted barriers are filled as AEC is summoned. Added cleanup of AEC on generate.

A hybrid soloution. May result in some more AECs in the world (> 10), but has slightly better command performance.

misode commented:
I'm almost happy with this PR.

Why did you move the `#cs:generate` function call to the end though?

ICY105 commented:
In case a user does something stupid (like create an infinite loop) then CS's spreading will still run. Otherwise the chunk would keep trying to generate over and over.

misode commented:
The way I see it: If a datapack messes up, the chunk should not be marked as generated. But now that I think about it, it makes more sense to run the tag afterwards.

I'm going to do some final testing in-game, and then I will merge.

ICY105 committed:
Replaced max distance limiter with predicate.

ICY105 commented:
I developed a predicate that detects if the chunk it's run from is next to an unloaded chunk. I replaced the max distance of 100 with this predicate check so all render distances are equally accounted for (though higher distances can leave more AECs in the world).

ICY105 committed:
Reverted the pre-fill change to the end (missed it :/)

ICY105 committed:
Improvement to multiplayer performance

misode commented on data/load/tags/functions/standard.json:
```diff
@@ -0,0 +1,5 @@
+{
+    "values": [
+		"cs:load"
+	]
```
Using this hook means that a data pack doesn't need to explicitly include this in their load tag. Because this is not a standard module pack, it should use the `#load:load` tag

misode commented on data/cs/functions/load.mcfunction:
```diff
@@ -1 +1,3 @@
 schedule function cs:main 2t replace
+#using vXX.XX.XX
+scoreboard players set $cs.ver load 010000
```
The convention is `id load` and using an incremental version. If you want to add a minor version, it should be in a separate fake player like `id_minor`.

I would use id `chunk_scan`.

misode commented on data/cs/functions/chunk_end/init_2.mcfunction:
```diff
@@ -0,0 +1,3 @@
+execute store result entity @s Pos[0] double 16 run data get entity @s Pos[0] 0.0625
+execute store result entity @s Pos[2] double 16 run data get entity @s Pos[2] 0.0625
+tag @s remove cs.init
```
`chunk_end/init` and `chunk_end/init2` are exectly the same as the `chunk` alternatives. I would remove these two functions and rename `chunk_end/generate` to `chunk/generate_end`.

misode commented on data/cs/functions/chunk/generate.mcfunction:
```diff
@@ -1,9 +1,13 @@
-fill ~ 0 ~ ~15 0 ~15 barrier replace bedrock
-kill @s
+execute store result entity @s Pos[0] double 16 run data get entity @s Pos[0] 0.0625
+execute store result entity @s Pos[2] double 16 run data get entity @s Pos[2] 0.0625
```
Why do you update the position when generating? This has already happened in the `init_2` function

ICY105 replied:
I was having some weird instances of the AECs getting displaced, which threw everything off. Idk what caused it, but I would guess Minecraft failed to update the position the first time around (maybe due to a lazy chunk?). This change prevents it from happening.

misode reviewed (commented):
Tested in game and functionality-wise, this works good! I'd like some changes related to load. Additionally I was considering to use the `chunk_scan` namespace to make it a bit less likely someone is going to have the same namespace.

misode commented:
I discovered you can use tnt to blow up the `moving_piston`'s at y=255 in the end. This allows players to regenerate chunks. If there is no fix for this, I would not include the end dimension in the scanning

ICY105 commented:
Hrm... I'm planning on adding End related world gen, so that's a problem. Maybe kill all tnt, creepers, end crystals, and ghast fireballs above y=250? (Think I that's all the explosion sources).

ICY105 committed:
Changed namespace to chunk_scan, changed load ID, combined init functions

ICY105 committed:
Fixed load tag

ICY105 committed:
Performance Improvements

misode commented:
I'm unsure of the last commit. I don't really want to mess with forceloading. This is not a pre-generator and it could remove already forceloaded chunks.

ICY105 commented:
I'm testing performance modifications. In practice it seems it's never actually used, so I'll just remove it (though it wouldn't remove forceloaded chunks as it only loads unloaded chunks for 1 tick to summon the entity).

ICY105 committed:
remove forceload

ICY105 committed:
oops

ICY105 committed:
Fix for entity spam in the end

ICY105 committed:
Added support for custom world types and dimensions

ICY105 committed:
cleanup, bump version

ICY105 committed:
fixes

ICY105 committed:
oops

ICY105 committed:
fixes

ICY105 committed:
Updated to support 1.17 snapshots

* Uses light blocks at world height limit (invulnerable in surivival, acts like air)
* Added support for any dimension world height
* Added support for custom dimensions (must be initialized manually)

ICY105 committed:
Revert "Updated to support 1.17 snapshots"

This reverts commit 9c9bb8504426c1d692ed608144d20d8179d55a87.

ICY105 committed:
update to entity only for 1.17-1.18 compat

uses markers, 1 per chunk
results in several hundred entities in the world
works in any dimension

ICY105 committed:
Implement 'super chunk markers'

Every 3x3 section of chunks gets a marker instead of every chunk. Results in x9 less entities in the world.

ICY105 committed:
generate function tag is now run at the bottom of the world instead of y=0

ICY105 committed:
Update LanternLoad, fix a typo

ICY105 committed:
change load -> load.status

ICY105 committed:
fix bug causes duplicate generation

ICY105 committed:
chunk scan actually fixed

ICY105 committed:
attempted fix... not working

ICY105 committed:
fix breaking if #generate caps command chain length

ICY105 committed:
add global ignore tags

ICY105 committed:
added smithed ignore tags to entities

ICY105 committed:
refactor with proper version control

ICY105 committed:
fix entity summon

ICY105 committed:
Update README.md

ICY105 committed:
Update pack.mcmeta

ICY105 committed:
Create dependencies.json

ICY105 committed:
Update README.md

ICY105 committed:
Update to 1.19

misode commented:
I will close this PR and simply let you maintain your fork
